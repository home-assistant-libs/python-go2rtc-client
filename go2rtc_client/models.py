"""Go2rtc Python models."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, Literal, TypeVar

from awesomeversion import AwesomeVersion
from mashumaro import field_options
from mashumaro.config import BaseConfig
from mashumaro.mixins.orjson import DataClassORJSONMixin
from mashumaro.types import SerializationStrategy


class _AwesomeVersionSerializer(SerializationStrategy):
    def serialize(self, value: AwesomeVersion) -> str:
        return str(value)

    def deserialize(self, value: str) -> AwesomeVersion:
        return AwesomeVersion(value)


_T = TypeVar("_T")


class _EmptyListInsteadNoneSerializer(SerializationStrategy, Generic[_T]):
    def serialize(self, value: list[_T]) -> list[_T]:
        return value

    def deserialize(self, value: list[_T] | None) -> list[_T]:
        if value is None:
            return []

        return value


@dataclass
class ApplicationInfo(DataClassORJSONMixin):
    """Application info model.

    Currently only the server version is exposed.
    """

    version: AwesomeVersion = field(
        metadata=field_options(serialization_strategy=_AwesomeVersionSerializer())
    )


@dataclass
class Streams(DataClassORJSONMixin):
    """Streams model."""

    streams: dict[str, Stream]


@dataclass
class Stream(DataClassORJSONMixin):
    """Stream model."""

    producers: list[Producer] = field(
        metadata=field_options(serialization_strategy=_EmptyListInsteadNoneSerializer())
    )


@dataclass
class Producer:
    """Producer model."""

    url: str
    media: list[str] = field(
        metadata=field_options(
            serialization_strategy=_EmptyListInsteadNoneSerializer(), alias="medias"
        )
    )

    class Config(BaseConfig):
        """Mashumaro config."""

        serialize_by_alias = True

    @classmethod
    def __pre_deserialize__(cls, d: dict[Any, Any]) -> dict[Any, Any]:
        """Pre deserialize."""
        # Ensure medias is always present
        d.setdefault("medias", [])
        return d


@dataclass
class WebRTCSdp(DataClassORJSONMixin):
    """WebRTC SDP model."""

    type: Literal["offer", "answer"]
    sdp: str


@dataclass
class WebRTCSdpOffer(WebRTCSdp):
    """WebRTC SDP offer model."""

    type: Literal["offer"] = field(default="offer", init=False)


@dataclass
class WebRTCSdpAnswer(WebRTCSdp):
    """WebRTC SDP answer model."""

    type: Literal["answer"] = field(default="answer", init=False)
