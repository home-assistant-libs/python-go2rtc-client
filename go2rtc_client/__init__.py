"""go2rtc client."""

from .rest import Go2RtcRestClient
from .models import Stream, WebRTCSdpAnswer, WebRTCSdpOffer
from . import ws

__all__ = ["Go2RtcRestClient", "Stream", "WebRTCSdpAnswer", "WebRTCSdpOffer", "ws"]
