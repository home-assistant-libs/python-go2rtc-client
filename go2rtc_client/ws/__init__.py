"""Websocket module."""

from .client import Go2RtcWsClient
from .messages import WebRTCCandidate, WebRTCOffer, WebRTCAnswer

__all__ = ["Go2RtcWsClient", "WebRTCCandidate", "WebRTCOffer", "WebRTCAnswer"]
