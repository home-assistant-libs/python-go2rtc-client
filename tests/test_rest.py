"""Asynchronous Python client for go2rtc."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from aiohttp.hdrs import METH_PUT
from awesomeversion import AwesomeVersion
import pytest

from go2rtc_client.models import WebRTCSdpOffer
from go2rtc_client.rest import _ApplicationClient, _StreamClient, _WebRTCClient
from tests import load_fixture

from . import URL

if TYPE_CHECKING:
    from aioresponses import aioresponses
    from syrupy import SnapshotAssertion

    from go2rtc_client import Go2RtcRestClient


async def test_application_info(
    responses: aioresponses,
    rest_client: Go2RtcRestClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test webrtc offer."""
    responses.get(
        f"{URL}{_ApplicationClient.PATH}",
        status=200,
        body=load_fixture("application_info_answer.json"),
    )
    resp = await rest_client.application.get_info()
    assert isinstance(resp.version, AwesomeVersion)
    assert resp == snapshot
    assert resp.to_dict() == snapshot


@pytest.mark.parametrize(
    "filename",
    [
        "streams_one.json",
        "streams_none.json",
    ],
    ids=[
        "one stream",
        "empty",
    ],
)
async def test_streams_get(
    responses: aioresponses,
    rest_client: Go2RtcRestClient,
    snapshot: SnapshotAssertion,
    filename: str,
) -> None:
    """Test get streams."""
    responses.get(
        f"{URL}{_StreamClient.PATH}",
        status=200,
        body=load_fixture(filename),
    )
    resp = await rest_client.streams.list()
    assert resp == snapshot


async def test_streams_add(
    responses: aioresponses,
    rest_client: Go2RtcRestClient,
) -> None:
    """Test add stream."""
    url = f"{URL}{_StreamClient.PATH}"
    params = {
        "name": "camera.12mp_fluent",
        "src": [
            "rtsp://test:test@192.168.10.105:554/Preview_06_sub",
            "ffmpeg:camera.12mp_fluent#audio=opus",
        ],
    }
    responses.put(
        url
        + "?name=camera.12mp_fluent"
        + "&src=ffmpeg%253Acamera.12mp_fluent%2523audio%253Dopus"
        + "&src=rtsp%253A%252F%252Ftest%253Atest%2540192.168.10.105%253A554%252F"
        + "Preview_06_sub",
        status=200,
    )
    await rest_client.streams.add(
        "camera.12mp_fluent", "rtsp://test:test@192.168.10.105:554/Preview_06_sub"
    )

    responses.assert_called_once_with(url, method=METH_PUT, params=params)


@pytest.mark.parametrize(
    ("server_version", "expected_result"),
    [
        ("0.0.0", False),
        ("1.9.3", False),
        ("1.9.4", True),
        ("1.9.5", True),
        ("2.0.0", False),
        ("BLAH", False),
    ],
)
async def test_version_supported(
    responses: aioresponses,
    rest_client: Go2RtcRestClient,
    server_version: str,
    expected_result: bool,
) -> None:
    """Test webrtc offer."""
    payload = json.loads(load_fixture("application_info_answer.json"))
    payload["version"] = server_version
    responses.get(
        f"{URL}{_ApplicationClient.PATH}",
        status=200,
        payload=payload,
    )
    assert await rest_client.validate_server_version() == expected_result


async def test_webrtc_offer(
    responses: aioresponses,
    rest_client: Go2RtcRestClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test webrtc offer."""
    camera = "camera.12mp_fluent"
    responses.post(
        f"{URL}{_WebRTCClient.PATH}?src={camera}",
        status=200,
        body=load_fixture("webrtc_answer.json"),
    )
    resp = await rest_client.webrtc.forward_whep_sdp_offer(
        camera,
        WebRTCSdpOffer("v=0..."),
    )
    assert resp == snapshot
