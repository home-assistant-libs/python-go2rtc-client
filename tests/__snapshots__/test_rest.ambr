# serializer version: 1
# name: test_streams_get[empty]
  dict({
  })
# ---
# name: test_streams_get[one stream]
  dict({
    'camera.12mp_fluent': dict({
      'producers': list([
        dict({
          'url': 'rtsp://test:test@192.168.10.105:554/Preview_06_sub',
        }),
      ]),
    }),
  })
# ---
# name: test_webrtc_offer
  dict({
    'sdp': 'v=0...',
    'type': 'answer',
  })
# ---
